File: src/main/java/dev/slimevr/poserecorder/PoseFrames.java
package dev.slimevr.poserecorder;

import dev.slimevr.vr.trackers.TrackerPosition;
import dev.slimevr.vr.trackers.TrackerUtils;
import io.eiren.util.collections.FastList;

import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;


public final class PoseFrames implements Iterable<TrackerFrame[]> {

	private final FastList<PoseFrameTracker> trackers;

	/**
	 * Creates a {@link PoseFrames} object with the provided list of
	 * {@link PoseFrameTracker}s as the internal {@link PoseFrameTracker} list
	 *
	 * @see {@link FastList}, {@link PoseFrameTracker}
	 */
	public PoseFrames(FastList<PoseFrameTracker> trackers) {
		this.trackers = trackers;
	}

	/**
	 * Creates a {@link PoseFrames} object with the specified initial tracker
	 * capacity
	 *
	 * @see {@link #PoseFrames(FastList)}
	 */
	public PoseFrames(int initialCapacity) {
		this.trackers = new FastList<PoseFrameTracker>(initialCapacity);
	}

	/**
	 * Creates a {@link PoseFrames} object with the default initial tracker
	 * capacity of {@code 5}
	 *
	 * @see {@link #PoseFrames(int)}
	 */
	public PoseFrames() {
		this(5);
	}

	/**
	 * Adds the provided {@link PoseFrameTracker} into the internal
	 * {@link PoseFrameTracker} list
	 *
	 * @return The {@link PoseFrameTracker} provided
	 * @see {@link List#add(Object)}, {@link PoseFrameTracker}
	 */
	public PoseFrameTracker addTracker(PoseFrameTracker tracker) {
		trackers.add(tracker);
		return tracker;
	}

	/**
	 * Removes the {@link PoseFrameTracker} at the specified index from the
	 * internal {@link PoseFrameTracker} list
	 *
	 * @return The {@link PoseFrameTracker} previously at the specified index
	 * @see {@link List#remove(int)}, {@link PoseFrameTracker}
	 */
	public PoseFrameTracker removeTracker(int index) {
		return trackers.remove(index);
	}

	/**
	 * Removes the specified {@link PoseFrameTracker} from the internal
	 * {@link PoseFrameTracker} list
	 *
	 * @return {@code true} if the internal {@link PoseFrameTracker} list
	 * contained the specified {@link PoseFrameTracker}
	 * @see {@link List#remove(Object)}, {@link PoseFrameTracker}
	 */
	public boolean removeTracker(PoseFrameTracker tracker) {
		return trackers.remove(tracker);
	}

	/**
	 * Clears the internal {@link PoseFrameTracker} list
	 *
	 * @see {@link List#clear()}, {@link PoseFrameTracker}
	 */
	public void clearTrackers() {
		trackers.clear();
	}

	/**
	 * Fake clears the internal {@link PoseFrameTracker} list by setting the
	 * size to zero
	 *
	 * @see {@link FastList#fakeClear()}, {@link PoseFrameTracker}
	 */
	public void fakeClearTrackers() {
		trackers.fakeClear();
	}

	/**
	 * @return The number of contained {@link PoseFrameTracker} objects
	 * @see {@link List#size()}, {@link PoseFrameTracker}
	 */
	public int getTrackerCount() {
		return trackers.size();
	}

	/**
	 * @return A list of the contained {@link PoseFrameTracker} objects
	 * @see {@link List}, {@link PoseFrameTracker}
	 */
	public List<PoseFrameTracker> getTrackers() {
		return trackers;
	}

	// #region Data Utilities

	/**
	 * A utility function to get the maximum Y value of the tracker associated
	 * with the {@link TrackerPosition#HMD} tracker position
	 *
	 * @return The maximum Y value of the tracker associated with the
	 * {@link TrackerPosition#HMD} tracker position
	 * @see {@link #getMaxHeight(TrackerPosition)}, {@link TrackerPosition#HMD}
	 */
	public float getMaxHmdHeight() {
		return getMaxHeight(TrackerPosition.HMD);
	}

	/**
	 * A utility function to get the maximum Y value of the tracker associated
	 * with the specified {@link TrackerPosition}
	 *
	 * @return The maximum Y value of the tracker associated with the specified
	 * {@link TrackerPosition}
	 * @see {@link TrackerPosition}
	 */
	public float getMaxHeight(TrackerPosition trackerPosition) {
		float maxHeight = 0f;

		PoseFrameTracker hmd = TrackerUtils
			.findNonComputedHumanPoseTrackerForBodyPosition(trackers, trackerPosition);

		if (hmd == null) {
			return maxHeight;
		}

		for (TrackerFrame frame : hmd) {
			if (frame.hasData(TrackerFrameData.POSITION) && frame.position.y > maxHeight) {
				maxHeight = frame.position.y;
			}
		}

		return maxHeight;
	}
	// #endregion

	/**
	 * @return The maximum number of {@link TrackerFrame}s contained within each
	 * {@link PoseFrameTracker} in the internal {@link PoseFrameTracker} list
	 * @see {@link PoseFrameTracker#getFrameCount()}, {@link PoseFrameTracker}
	 */
	public int getMaxFrameCount() {
		int maxFrames = 0;

		for (PoseFrameTracker tracker : trackers) {
			if (tracker != null && tracker.getFrameCount() > maxFrames) {
				maxFrames = tracker.getFrameCount();
			}
		}

		return maxFrames;
	}

	/**
	 * Using the provided array buffer, get the {@link TrackerFrame}s contained
	 * within each {@link PoseFrameTracker} in the internal
	 * {@link PoseFrameTracker} list at the specified index
	 *
	 * @return The number of frames written to the buffer
	 * @see {@link PoseFrameTracker#safeGetFrame(int)}, {@link TrackerFrame},
	 * {@link PoseFrameTracker}
	 */
	public int getFrames(int frameIndex, TrackerFrame[] buffer) {
		int frameCount = 0;

		for (PoseFrameTracker tracker : trackers) {
			if (tracker == null) {
				continue;
			}

			TrackerFrame frame = tracker.safeGetFrame(frameIndex);

			if (frame == null) {
				continue;
			}

			buffer[frameCount++] = frame;
		}

		return frameCount;
	}

	/**
	 * Using the provided {@link List} buffer, get the {@link TrackerFrame}s
	 * contained within each {@link PoseFrameTracker} in the internal
	 * {@link PoseFrameTracker} list at the specified index
	 *
	 * @return The number of frames written to the buffer
	 * @see {@link PoseFrameTracker#safeGetFrame(int)},
	 * {@link List#set(int, Object)}, {@link TrackerFrame},
	 * {@link PoseFrameTracker}
	 */
	public int getFrames(int frameIndex, List<TrackerFrame> buffer) {
		int frameCount = 0;

		for (PoseFrameTracker tracker : trackers) {
			if (tracker == null) {
				continue;
			}

			TrackerFrame frame = tracker.safeGetFrame(frameIndex);

			if (frame == null) {
				continue;
			}

			buffer.set(frameCount++, frame);
		}

		return frameCount;
	}

	/**
	 * @return The {@link TrackerFrame}s contained within each
	 * {@link PoseFrameTracker} in the internal {@link PoseFrameTracker} list at
	 * the specified index
	 * @see {@link PoseFrameTracker#safeGetFrame(int)}, {@link TrackerFrame},
	 * {@link PoseFrameTracker}
	 */
	public TrackerFrame[] getFrames(int frameIndex) {
		TrackerFrame[] trackerFrames = new TrackerFrame[trackers.size()];
		getFrames(frameIndex, trackerFrames);
		return trackerFrames;
	}

	@Override
	public Iterator<TrackerFrame[]> iterator() {
		return new PoseFrameIterator(this);
	}

	public class PoseFrameIterator implements Iterator<TrackerFrame[]> {

		private final PoseFrames poseFrame;
		private final TrackerFrame[] trackerFrameBuffer;

		private int cursor = 0;

		public PoseFrameIterator(PoseFrames poseFrame) {
			this.poseFrame = poseFrame;
			trackerFrameBuffer = new TrackerFrame[poseFrame.getTrackerCount()];
		}

		@Override
		public boolean hasNext() {
			if (trackers.isEmpty()) {
				return false;
			}

			for (PoseFrameTracker tracker : trackers) {
				if (tracker != null && cursor < tracker.getFrameCount()) {
					return true;
				}
			}

			return false;
		}

		@Override
		public TrackerFrame[] next() {
			if (!hasNext()) {
				throw new NoSuchElementException();
			}

			poseFrame.getFrames(cursor++, trackerFrameBuffer);

			return trackerFrameBuffer;
		}
	}
}

File: server/core/src/main/java/dev/slimevr/poseframeformat/PoseFrames.kt
package dev.slimevr.poseframeformat

import dev.slimevr.poseframeformat.trackerdata.TrackerFrame
import dev.slimevr.poseframeformat.trackerdata.TrackerFrames
import dev.slimevr.tracking.trackers.TrackerPosition
import io.eiren.util.collections.FastList

class PoseFrames : Iterable<Array<TrackerFrame?>> {
	val frameHolders: FastList<TrackerFrames>

	/**
	 * Frame interval in seconds
	 */
	var frameInterval: Float = 0.02f

	/**
	 * Creates a [PoseFrames] object with the provided list of
	 * [TrackerFrames]s as the internal [TrackerFrames] list.
	 *
	 * @see [FastList]
	 * @see [TrackerFrames]
	 */
	constructor(frameHolders: FastList<TrackerFrames>) {
		this.frameHolders = frameHolders
	}

	/**
	 * Creates a [PoseFrames] object with the specified initial tracker
	 * capacity.
	 *
	 * @see [PoseFrames]
	 */
	constructor(initialCapacity: Int = 5) {
		frameHolders = FastList(initialCapacity)
	}

	/**
	 * @return The [TrackerFrames] associated with [position] at frame
	 * index [index].
	 */
	fun getTrackerForPosition(position: TrackerPosition, index: Int = 0): TrackerFrames? {
		for (tracker in frameHolders) {
			if (tracker.tryGetFrame(index)?.trackerPosition == position) return tracker
		}
		return null
	}

	// region Data Utilities
	/**
	 * @return The maximum Y value of the [TrackerFrames] associated with the
	 * [TrackerPosition.HEAD] tracker position on the first frame, otherwise `0f` if
	 * no [TrackerFrames] is associated with [TrackerPosition.HEAD] or if there are no
	 * valid positions.
	 * @see [getMaxHeight]
	 */
	val maxHmdHeight: Float
		get() {
			return getMaxHeight(
				getTrackerForPosition(TrackerPosition.HEAD)
					?: return 0f,
			)
		}

	/**
	 * @return The maximum Y value of the [trackerFrames], otherwise `0f` if
	 * there are no valid positions.
	 * @see [TrackerPosition]
	 */
	fun getMaxHeight(trackerFrames: TrackerFrames): Float {
		var maxHeight = 0f
		for (frame in trackerFrames.frames) {
			val framePosition = frame?.tryGetPosition() ?: continue

			if (framePosition.y > maxHeight) {
				maxHeight = framePosition.y
			}
		}

		return maxHeight
	}
	// endregion

	/**
	 * @return The maximum number of [TrackerFrame]s contained within each
	 * [TrackerFrames] in the internal [TrackerFrames] list.
	 * @see [TrackerFrames.frames]
	 * @see [List.size]
	 */
	val maxFrameCount: Int
		get() {
			return frameHolders.maxOfOrNull { tracker -> tracker.frames.size } ?: 0
		}

	/**
	 * Using the provided array buffer, get the [TrackerFrame]s contained
	 * within each [TrackerFrames] in the internal
	 * [TrackerFrames] list at the specified index.
	 *
	 * @return The number of frames written to the buffer.
	 * @see [TrackerFrames.tryGetFrame]
	 */
	fun getFrames(frameIndex: Int, buffer: Array<TrackerFrame?>): Int {
		var frameCount = 0
		for (tracker in frameHolders) {
			if (tracker == null) {
				continue
			}
			val frame = tracker.tryGetFrame(frameIndex) ?: continue
			buffer[frameCount++] = frame
		}
		return frameCount
	}

	/**
	 * Using the provided [List] buffer, get the [TrackerFrame]s
	 * contained within each [TrackerFrames] in the internal
	 * [TrackerFrames] list at the specified index.
	 *
	 * @return The number of frames written to the buffer.
	 * @see [TrackerFrames.tryGetFrame]
	 */
	fun getFrames(frameIndex: Int, buffer: MutableList<TrackerFrame?>): Int {
		var frameCount = 0
		for (tracker in frameHolders) {
			if (tracker == null) {
				continue
			}
			val frame = tracker.tryGetFrame(frameIndex) ?: continue
			buffer[frameCount++] = frame
		}
		return frameCount
	}

	/**
	 * @return The [TrackerFrame]s contained within each
	 * [TrackerFrames] in the internal [TrackerFrames] list at
	 * the specified index.
	 * @see [TrackerFrames.tryGetFrame]
	 */
	fun getFrames(frameIndex: Int): Array<TrackerFrame?> {
		val trackerFrames = arrayOfNulls<TrackerFrame>(frameHolders.size)
		getFrames(frameIndex, trackerFrames)
		return trackerFrames
	}

	override fun iterator(): Iterator<Array<TrackerFrame?>> = PoseFrameIterator(this)

	inner class PoseFrameIterator(private val poseFrame: PoseFrames) : Iterator<Array<TrackerFrame?>> {
		private val trackerFrameBuffer: Array<TrackerFrame?> = arrayOfNulls(poseFrame.frameHolders.size)
		private val maxCursor = poseFrame.maxFrameCount
		private var cursor = 0

		override fun hasNext(): Boolean = frameHolders.isNotEmpty() && cursor < maxCursor

		override fun next(): Array<TrackerFrame?> {
			if (!hasNext()) {
				throw NoSuchElementException()
			}
			poseFrame.getFrames(cursor++, trackerFrameBuffer)
			return trackerFrameBuffer
		}
	}
}
